src/request.ts

```diff
@@ -0,0 +1,145 @@
+import {isObject} from 'lodash';
+import 'isomorphic-fetch';
+
+/**
+ * Return a promise that is resolved or rejected depending on the response's
+ * status code.
+ */
+export const checkStatus = (res: Response): Promise<Response> => {
+    if (res.status >= 300) {
+        return Promise.reject(res);
+    }
+    return Promise.resolve(res);
+};
+
+/**
+ * Calls fetch on provided url with default options necessary for interacting
+ * with our JSON API. Parses the JSON, provides appropriate headers, and asserts
+ * a valid status from the server.
+ */
+export const fetchJSON = (
+    url: string,
+    {headers, method, mode, ...options}: RequestInit = {}
+): Promise<{}> => {
+    let fetchHeaders = headers as HeadersInit;
+
+    if (method !== 'GET') {
+        fetchHeaders = {
+            'Content-Type': 'application/json',
+            ...headers,
+        };
+    }
+
+    const fetchOptions = {
+        method,
+        mode,
+        headers: fetchHeaders,
+        credentials: mode === 'cors' ? 'include' : 'same-origin',
+        ...options,
+    } as RequestInit;
+
+    return fetch(url, fetchOptions)
+        .then(checkStatus)
+        .then((res: Response) => {
+            let resJSON = {};
+
+            if ('json' in res) {
+                resJSON = res.json();
+            }
+
+            return resJSON;
+        });
+};
+
+export interface ArgumentErrors {
+    [key: string]: [string];
+}
+
+export interface ParsedResponseError {
+    error: string;
+    description: string;
+    argumentErrors: ArgumentErrors;
+}
+
+export interface JSONResponseData {
+    error?: string;
+    error_description?: string;
+    error_detail?: {
+        ARGUMENTS_ERROR?: ArgumentErrors;
+        [propName: string]: any;
+    };
+}
+
```

Comment:
what do you think about creating a separate file for type definitions?

Reply:
Makes sense, but I'll have to see how it works. I wouldn't want to have to import from that separate file and then export again.

Reply:
Seemed to work ok!